server/src/main/java/org/cloudfoundry/identity/uaa/oauth/token/UserTokenGranter.java

```diff
@@ -97,6 +104,16 @@ protected Authentication validateRequest(TokenRequest request) {
         ClientDetails receiving = clientDetailsService.loadClientByClientId(request.getRequestParameters().get(CLIENT_ID), IdentityZoneHolder.get().getId());
         super.validateGrantType(GRANT_TYPE_REFRESH_TOKEN, receiving);
 
+        //7. receiving client must have a client secret
+        try {
+            UsernamePasswordAuthenticationToken clientAuth = 
+                    new UsernamePasswordAuthenticationToken(request.getRequestParameters().get(CLIENT_ID), "");
+            clientAuthenticationProvider.authenticate(clientAuth);
+            throw new InvalidClientException("Grant type is not allowed for this client");
```

Comment:
This code is very difficult to understand. It seems that an exception is expected in order for the grant to be validated. Could this (and perhaps the authenticate method that's called here) be restructured so that the requirement this is validating is easier to understand from looking at the code?